### ISSUE TITLE
Support AWS SDK 2.0

### ISSUE THREAD

**Issue by mv78** [spring-cloud/spring-cloud-aws#267]
AWS currently in the process or releasing SDK 2.0. It would be good to start integrating it into the project as it progresses to GA.

**Comment by aemruli** [comment]
@mv78 just checked today. At least from the information I found they are in the preview. Hence we are waiting once AWS has released a final 2.0. 
I have updated the project to use the latest AWS SDK 1.x

**Comment by ajainy** [comment]
Can we upgrade AWS SDK v1.11.321 (latest as of May, 2018) ..? 
@aemruli 

Or please share your last commit, i will do similar.

**Comment by aemruli** [comment]
@ajainy we have upgraded master to 1.11.336

**Comment by michl-b** [comment]
Are there any news on this? 
Usage with JDK11 gives some warnings on a class used in the old V1 of aws-sdk-java

`
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.amazonaws.util.XpathUtils (file:/Users/michael/.m2/repository/com/amazonaws/aws-java-sdk-core/1.11.415/aws-java-sdk-core-1.11.415.jar) to method com.sun.org.apache.xpath.internal.XPathContext.getDTMManager()
WARNING: Please consider reporting this to the maintainers of com.amazonaws.util.XpathUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
`

**Comment by kmualem** [comment]
Hi guys

Any news about AWS SDK version 2?

**Comment by michaelwiles** [comment]
Yes, problem with AWS SDK version 1 is it doesn't play nice with JDK 9 as it duplicates packages across individual jars.

Although an upgrade like this would touch a lot of classes it wouldn't be too hard to do as it's a simple job to upgrade to the new SDK and the new SDK is a lot easier to work with (imo it has a more fluent api).

**Comment by artembilan** [comment]
See another reason to do this: https://github.com/spring-cloud/spring-cloud-aws/issues/491
Also here: https://github.com/spring-projects/spring-integration-aws/issues/155

Thanks

**Comment by tinexw** [comment]
I started working on this here https://github.com/tinexw/spring-cloud-aws/commit/c622bef7482aacd0fc49a212e1fe1476bed64f57

If anyone wants to join, please let me know!

**Comment by spencergibb** [comment]
@tinexw not sure how this would work as updating major versions of a dependency would be a breathing change

**Comment by tinexw** [comment]
Yeah, that's why I was asking what the preferred approach in this case is. Release a new major version?

**Comment by OlgaMaciaszek** [comment]
We would be very happy to take this upgrade PR, but we can't put backwards-incompatible changes within the Hoxton release train, cause it's not going to be a major. We will be shortly deciding if our next release train will be major (we will probably branch it out in a few months). If yes, then these changes could be merged there.

**Comment by tinexw** [comment]
Ok, thanks for the update. There is quite a lot left. I will pick it up again once you know for sure if it’s going to be a major release.

**Comment by sp00m** [comment]
Yet another reason to upgrade: [EKS supports 2.10.11+ only](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-minimum-sdk.html)... Any ETA yet?

**Comment by tcheek-bw** [comment]
Is this still being worked on? The 2.0 SDK would be really nice to have.

**Comment by spencergibb** [comment]
@tinexw started working on it and now that master would be a major release we would be happy to accept it.

**Comment by tinexw** [comment]
Good to know :-) I will pick it back up. Would be great to get some support especially in regards to testing later on.

**Comment by tinexw** [comment]
So... quick update. I did the first "stupid" part of upgrading all the libraries and fixing all compile error according to the [changelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md#5-profile-file-changes). 

There is still a lot to do (s. list below) but before I continue I would like to discuss some potential blockers:

**1. AmazonSQSBufferedAsyncClient not supported**
The [AmazonSQSBufferedAsyncClient](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/sqs/buffered/AmazonSQSBufferedAsyncClient.html) is no longer supported according to the [AWS docs](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-client-side-buffering-request-batching.html):

> The AWS SDK for Java 2.x isn't currently compatible with the AmazonSQSBufferedAsyncClient. 

There already is a ticket in the AWS SDK GitHub but it's still open: https://github.com/aws/aws-sdk-java-v2/issues/848 resp. https://github.com/aws/aws-sdk-java-v2/issues/848 

**2. No common interface for Sync and Async client**
The 1.x SDK had a common interface for the sync and async client. This is no longer the case. 

Previously, the user could add either the async or the sync client which would prevent a client from being autoconfigured by the framework. Additionally, a custom client could be passed to certain framework classes, e.g. `NotificationMessagingTemplate`.

I would suggest to
- Always autoconfigure the sync client and if someone wants to use the async client, they can add it themselves. 
- In the first step, only accept one client (either sync or async - depending on which methods are used) in all customizable classes. As a next step, the classes could be duplicated to accept both clients. 

**3. Messaging: Sync and Async client used**
In the 1.x SDK it was possible to call sync and async methods from the async client. The messaging module makes heavy use of this. I see the following options with the new SDK:
- Keep functionality the same by adding both clients and using the respective one (preferred by me)
- Use async methods everywhere
- Use sync methods everywhere

Please let me know what you think @spencergibb @OlgaMaciaszek 


Things to do:
- [x] Upgrade libraries and fix compile errors
- [ ] Discuss and resolve potential blockers
- [ ] Resolve remaining TODOs in code (non straightforward compile errors)
- [ ] Run and fix all unit tests
- [ ] Integration Tests: Upgrade libraries, fix, run and fix all issues
- [ ] Manual testing of everything not covered by integration tests
- [ ] Update documentation

**Comment by sp00m** [comment]
Just as an FYI, this ticket blocked us going forward, so we decided to develop our own abstraction for our Spring WebFlux projects, so that we could connect to SQS from within EKS: https://github.com/qudini/qudini-reactive/tree/master/qudini-reactive-sqs. Very SQS (and WebFlux) focused obviously, and by no means intended to replace https://github.com/spring-cloud/spring-cloud-aws, but maybe it can help resolving this issue somehow :)

**Comment by maciejwalkowiak** [comment]
Thanks @sp00m! We are going to move to SDK v2 in 3.0 and revisit SQS integration so we are definitely going to look at your repository for an inspiration.

**Comment by internetstaff** [comment]
Perhaps Spring Cloud AWS 3.0 could just be an upgrade to SDK v2 in order to knock that out sooner rather than later?

We're starting to hit issues like SSO that simply aren't supported in SDK v1.

**Comment by spencergibb** [comment]
The repo that would handle this is https://github.com/awspring/spring-cloud-aws
